src/hotspot/share/memory/metaspace/chunkManager.cpp

```diff
@@ -279,8 +296,22 @@ void ChunkManager::return_chunk_locked(Metachunk* c) {
 //
 // On success, true is returned, false otherwise.
 bool ChunkManager::attempt_enlarge_chunk(Metachunk* c) {
-  MutexLocker fcl(Metaspace_lock, Mutex::_no_safepoint_check_flag);
-  return c->vsnode()->attempt_enlarge_chunk(c, &_chunks);
+  bool enlarged;
+  chunklevel_t old_level;
+
+  {
+    MutexLocker fcl(Metaspace_lock, Mutex::_no_safepoint_check_flag);
+    old_level = c->level();
+    enlarged = c->vsnode()->attempt_enlarge_chunk(c, &_chunks);
+  }
+
+  if (enlarged) {
```

Comment:
What's the reason of calling the ASAN functions outside of the lock?

Will we have a race condition if two threads call ChunkManager::attempt_enlarge_chunk() at the same time and get to here out of order? @tstuefe

Reply:
To be enlarged chunk belongs to the class loader already, and we are inside the class loader metaspace lock.

Reply:
If there's no particular reason to do it outside of the lock, I would prefer to have the ASAN call while holding the lock. That way you don't need to worry about race conditions.

```
MutexLocker fcl(Metaspace_lock, Mutex::_no_safepoint_check_flag);
old_level = c->level();
enlarged = c->vsnode()->attempt_enlarge_chunk(c, &_chunks);
ASAN_UNPOISON_MEMORY_REGION(....);
```

Reply:
Sure, why not. The reason would be reducing time inside the lock, but I'm unemotional. Probably does not matter much.

Reply:
Poisoning isn't free, it requires memset over a chunk of memory in the shadow region. This will almost guarantee a cache miss. The idea was doing it outside the lock to help speed things up, as with ASan build some tests timeout.